promoter/promoter.go

```diff
@@ -164,3 +182,71 @@ func (p *Promoter) staticAddrDiff(ctx context.Context) (toAdd []WatchedAddress,
 	}
 	return
 }
+
+// threadedCreditTransactions continuously polls the db for uncreditted txns and
+// notifies the credit promoter about them.
+func (p *Promoter) threadedCreditTransactions() {
+	// TODO: Fetch all txns with credited == false.
+
+	// TODO: Credit the promoter. NOTE: This needs to be idempotent so we
+	// include a txn ID in the request. Requesting a credit with the same
+	// txn ID should still only credit it once.
+
+	// TODO: Update the credited txns in the database as credited = true.
+}
+
+// threadedPollTransactions continuously polls skyd for transactions related to
+// watched addresses and writes them to the DB.
+func (p *Promoter) threadedPollTransactions() {
+	t := time.NewTicker(txnPolInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-p.staticBGCtx.Done():
+			return
+		case <-t.C:
+		}
+		p.staticLogger.WithTime(time.Now()).Info("Starting to poll transactions from skyd")
+
+		// Get used addresses.
+		c, err := p.staticColWatchedAddresses().Find(p.staticBGCtx, bson.M{
+			"$or": bson.A{
+				bson.M{"user_id": bson.M{"$exists": true}},
+				bson.M{"user_id": bson.M{"$ne": ""}},
+			},
+		})
+		if err != nil {
+			p.staticLogger.WithError(err).Error("Failed to fetch used addresses")
+			continue
+		}
+
+		// For each one get the related txns from skyd and save them to
+		// the db.
+		var nAddresssInserted, nTxnsInserted int
+		for c.Next(p.staticBGCtx) {
+			// Get address.
+			var wa WatchedAddress
+			if err := c.Decode(&wa); err != nil {
+				p.staticLogger.WithError(err).Error("Failed to decode address")
+				continue // try next address
+			}
+
+			// Fetch related txns from skyd.
+			txns, err := p.staticTxnsByAddress(wa.Address)
```

Comment:
Here we're making an external call while keeping a DB cursor open. That's not great because the results are being kept in mongo's memory. It would be better to grab all addresses first (they won't be that many and they won't strain our RAM too much) and then process them.